File: 프로그래머스/Lv.3/42628. 이중우선순위큐/이중우선순위큐.java
import java.util.*;
class Solution {
    public int[] solution(String[] operations) {
        PriorityQueue<Integer> priorityQueueMin = new PriorityQueue<>();
        PriorityQueue<Integer> priorityQueueMax = new PriorityQueue<>(Collections.reverseOrder());

        Arrays.stream(operations)
            .map(str -> str.split(" "))
            .forEach(strings ->
                     {
                         if (strings[0].equals("I")) {
                             priorityQueueMax.offer(Integer.parseInt(strings[1]));
                             priorityQueueMin.offer(Integer.parseInt(strings[1]));
                         } else if (strings[0].equals("D") && strings[1].equals("1")) {  //최대값 삭제
                             priorityQueueMin.remove(priorityQueueMax.poll());
                         } else if (strings[0].equals("D") && strings[1].equals("-1")) { //최소값 삭제
                             priorityQueueMax.remove(priorityQueueMin.poll());
                         }
                     });

        List<Integer> list1 = new ArrayList<>(priorityQueueMin);
        List<Integer> list2 = new ArrayList<>(priorityQueueMax);
        list1.retainAll(list2);

        if (list1.isEmpty()) {
            return new int[]{0,0};
        }

        list1.sort(Collections.reverseOrder());
        return new int[]{list1.get(0),list1.get(list1.size()-1)};
    }
}

File: 프로그래머스/Lv.3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    min_heap = []
    max_heap = []

    for operation in operations:
        op, num = operation.split()
        num = int(num)

        if op == "I":
            heapq.heappush(min_heap, num)
            heapq.heappush(max_heap, -num)
        elif op == "D" and num == 1:
            if max_heap:
                min_heap.remove(-heapq.heappop(max_heap))
        elif op == "D" and num == -1:
            if min_heap:
                max_heap.remove(-heapq.heappop(min_heap))

    common_values = list(set(min_heap) & set(-x for x in max_heap))

    if not common_values:
        return [0, 0]

    common_values.sort(reverse=True)
    return [common_values[0], common_values[-1]]